### ISSUE TITLE
weird training process&result during training yolo11 with 8.3.28

### ISSUE THREAD

**Issue by cvtower** [ultralytics/ultralytics#18058]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

whatever changes made to the source code, the training log、result(all loss、all map values) always the same, and the only difference is training time and time stamp.
additional info：
e.g. with custom dataset v1, train yolo11m i get a result, then train  yolo11s/l get another different result.
however, i just add private opts or keep the same source code, and then traing yolo11m again, the result  totally same with previous m version.
i tried to remove label .cache file or remove private opt codes and reinstall ultralytics, but this issue still exists. 

<img width="1550" alt="截屏2024-12-06 11 39 12" src="https://github.com/user-attachments/assets/e29e72bc-b519-4617-98b8-dcf695e8fddb">

<img width="1573" alt="截屏2024-12-06 11 38 57" src="https://github.com/user-attachments/assets/fc9012ae-dc3d-4446-bce0-01600ec34e56">

### Environment

ultralytics-8.3.28
ubuntu
pytorch 2.3.1+cu118
python 3.10

### Minimal Reproducible Example

official traing cli

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cvtower, thank you for bringing this to our attention and for using Ultralytics 🚀! We appreciate your detailed report and the included screenshots—it’s always helpful to have such context.

To help us debug this 🐛 bug further, could you share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/)? This will allow us to investigate the issue more efficiently. You mentioned using the "official training CLI," but if there are any specific arguments or modifications you're applying, please include those details as well.

In the meantime, please ensure that your environment is up-to-date by upgrading your `ultralytics` package, along with the necessary dependencies, to the latest version:

```bash
pip install -U ultralytics
```

You can also verify that the issue is not related to your current environment. YOLO can be run in a variety of verified environments, such as:
- **Notebooks** with free GPUs: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For additional debugging tips, check out our [Docs](https://docs.ultralytics.com), particularly the sections on [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/). You may also find relevant guidance in the [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) section to ensure optimal results.

If you'd like to connect with our community for support or discussion:
- Join the conversation on [Discord](https://discord.com/invite/ultralytics) 🎧
- Post on our [Discourse forum](https://community.ultralytics.com) for in-depth discussions
- Visit our [Subreddit](https://reddit.com/r/ultralytics) to share knowledge and experiences

Lastly, please note that this is an automated response 🤖, and an Ultralytics engineer will review your issue and provide further assistance soon. Thank you for your patience and for helping us improve!

**Comment by Y-T-G** [comment]
Is this present in the latest version?

**Comment by cvtower** [comment]
> Is this present in the latest version?

no, with 8.3.28.
i have produced 20+ version models for different tasks with 8.3.28 last month, but did not notice this problem.

**Comment by Y-T-G** [comment]
Please report it as a bug only if exists in the latest version.

**Comment by Laughing-q** [comment]
@cvtower Hey thanks for reporting! But I don't get where's the problem of this behavior...this is what we expected actually, to get deterministic results in different trainings(with same training settings).
We have an argument here `deterministic` and set to `True` by default.https://github.com/ultralytics/ultralytics/blob/c284a9e9744afb867c1acb6518d65cb21b7a1e75/ultralytics/cfg/default.yaml#L27
and if you want different results for your each training, you can try to set this argument to `False`.

**Comment by cvtower** [comment]
> @cvtower Hey thanks for reporting! But I don't get where's the problem of this behavior...this is what we expected actually, to get deterministic results in different trainings(with same training settings). We have an argument here `deterministic` and set to `True` by default.
> 
> https://github.com/ultralytics/ultralytics/blob/c284a9e9744afb867c1acb6518d65cb21b7a1e75/ultralytics/cfg/default.yaml#L27
> 
> 
> and if you want different results for your each training, you can try to set this argument to `False`.

thanks very much for ur reply.
i will try it and report later.

**Comment by cvtower** [comment]
> @cvtower Hey thanks for reporting! But I don't get where's the problem of this behavior...this is what we expected actually, to get deterministic results in different trainings(with same training settings). We have an argument here `deterministic` and set to `True` by default.
> 
> https://github.com/ultralytics/ultralytics/blob/c284a9e9744afb867c1acb6518d65cb21b7a1e75/ultralytics/cfg/default.yaml#L27
> 
> 
> and if you want different results for your each training, you can try to set this argument to `False`.



Hello,

i just test the option change and the training logs(loss,acc every epoch) r totally same with previous version except training time cost.
With a certain dataset, every time we trained the model with different optimizations(data aug implements and other hyper-paras), and got same training logs.
Is this the expected result? any tips would be appreciated.


ultralytics/cfg/default.yaml+30:
deterministic: False # (bool) whether to enable deterministic mode

on the left is the previous version of model last week, the right one is new model few minutes before. 
![image](https://github.com/user-attachments/assets/19c5dcb9-6ef7-4175-87ef-a2a18d6eca85)

**Comment by Y-T-G** [comment]
@cvtower What's the training command you're using?

**Comment by Y-T-G** [comment]
To set `deterministic=False` you just pass it as an argument to `model.train()`. Same for all the arguments inside `default.yaml`

**Comment by cvtower** [comment]
> @cvtower What's the training command you're using?

I just re-pull the ultralytics-8.3.47 offical code, reinstall ultralytics locally and test a custom dataset with the training cli:

yolo task=detect train data=****yaml  model=yolo11m.yaml pretrained=./yolo11m.pt  device=0,1  epochs=10 imgsz=640 batch=128 flipud=0.5   >****_test.log
yolo task=detect train data=****yaml  model=yolo11m.yaml pretrained=./yolo11m.pt  device=0,1  epochs=10 imgsz=640 batch=128 flipud=0.5   >****_test1.log


and the results is:
![image](https://github.com/user-attachments/assets/5e119afe-6d03-49d4-acc5-e94f1c9a45bf)

**Comment by Laughing-q** [comment]
@cvtower Hey I'm a bit confused here actually, as you mentioned that you were launching trainings in different training settings.
> With a certain dataset, every time we trained the model with different optimizations(data aug implements and other hyper-paras), and got same training logs.

But from your commands you were using the same commands for two training, which is expected to get exactly the same results.
If you want these trainings are different everytime, you could:
1. set `deterministic=False` as mentioned
2. use different `seed`, for instance you train one training within `seed=1`, then use `seed=2` for another training.

**Comment by cvtower** [comment]
> @cvtower Hey I'm a bit confused here actually, as you mentioned that you were launching trainings in different training settings.
> 
> > With a certain dataset, every time we trained the model with different optimizations(data aug implements and other hyper-paras), and got same training logs.
> 
> But from your commands you were using the same commands for two training, which is expected to get exactly the same results. If you want these trainings are different everytime, you could:
> 
> 1. set `deterministic=False` as mentioned
> 2. use different `seed`, for instance you train one training within `seed=1`, then use `seed=2` for another training.

Thank u ever so muchhh.

In fact, i just enabled the A.RandomRotate90() aug within the augment.py file, and modified aug paras, but found no changes in the training log.
Have tried to set `deterministic=False` and no changes found.
i will try alternative seeds.

**Comment by Laughing-q** [comment]
@cvtower 
> A.RandomRotate90()

which is from albumentations? then please make sure you have it installed as well, as in our code we directly skip this part(without throwing any errors) if it;s not installed.
Also since you are modifying the source code please make sure you are modifying the right please, I remember some users before they git cloned the repo and modified the source code and it did not work because they are actually using the `ultralytics` package installed from pypi rather than the cloned one.

**Comment by cvtower** [comment]
> @cvtower Hey I'm a bit confused here actually, as you mentioned that you were launching trainings in different training settings.
> 
> > With a certain dataset, every time we trained the model with different optimizations(data aug implements and other hyper-paras), and got same training logs.
> 
> But from your commands you were using the same commands for two training, which is expected to get exactly the same results. If you want these trainings are different everytime, you could:
> 
> 1. set `deterministic=False` as mentioned
> 2. use different `seed`, for instance you train one training within `seed=1`, then use `seed=2` for another training.

with alternative seed, training log changes.

yeah, thanks again for ur help.

**Comment by Laughing-q** [comment]
@cvtower Nice! I'm going to close the issue then, please free feel to reopen it if you have related questions! Thanks!

**Comment by cvtower** [comment]
> @cvtower
> 
> > A.RandomRotate90()
> 
> which is from albumentations? then please make sure you have it installed as well, as in our code we directly skip this part(without throwing any errors) if it;s not installed. Also since you are modifying the source code please make sure you are modifying the right please, I remember some users before they git cloned the repo and modified the source code and it did not work because they are actually using the `ultralytics` package installed from pypi rather than the cloned one.

yes, i get it.
i do reinstall the ultralytics lib locally every time after private modification.

Thanks again for you help. @Y-T-G @Laughing-q

**Comment by Y-T-G** [comment]
You can just do `pip install -e .` (dot at the end)

That way any changes you make locally is applied without reinstallation.
